File: Programmers_Python/lv2_이모티콘 할인행사.py
data = [10, 20, 30, 40]
discount = []
ans = [0,0]
def dfs(tmp, depth):
    if depth==len(tmp):
        discount.append(tmp[:])
        return

    for d in data:
        tmp[depth] += d
        dfs(tmp, depth+1)
        tmp[depth] -= d


def solution(users, emoticons):

    dfs([0]*len(emoticons), 0)

    for i in range(len(discount)):
        join, price = 0, [0]*len(users)
        for j in range(len(emoticons)):
            for k in range(len(users)):
                if users[k][0] <= discount[i][j]:
                    price[k] += emoticons[j] * (100-discount[i][j]) / 100

        for u in range(len(users)):
            if price[u] >= users[u][1]:
                join+=1
                price[u] = 0

        if join >= ans[0]:
            if join == ans[0]:
                ans[1] = max(ans[1], sum(price))
            else:
                ans[1] = sum(price)
            ans[0] = join

    return ans


users = [[40, 10000], [25, 10000]]
emoticons = [7000, 9000]

print(solution(users, emoticons))

File: Programmers_Java/Programmers_Java/src/lv2_이모티콘할인행사.java
import java.util.ArrayList;
import java.util.Arrays;

public class lv2_이모티콘할인행사 {
    static ArrayList<int[]> discount = new ArrayList<>();
    static int[] data = {10, 20, 30, 40};
    static int[] tmp;
    static int[] ans = new int[2];
    public static int[] solution(int[][] users, int[] emoticons) {
        tmp = new int[emoticons.length];
        dfs(0);

        for (int i = 0; i < discount.size(); i++) {
            int join = 0;
            int[] price = new int[users.length];
            for (int j = 0; j < emoticons.length; j++) {
                for (int k = 0; k < users.length; k++) {
                    if (users[k][0] <= discount.get(i)[j]) {
                        price[k] += emoticons[j]*(100- discount.get(i)[j])/100;
                    }
                }
            }

            for (int k = 0; k < users.length; k++) {
                if (price[k] >= users[k][1]) {
                    join+=1;
                    price[k] = 0;
                }
            }
            if (join >= ans[0]) {
                if (join==ans[0]){
                    ans[1] = Math.max(ans[1], Arrays.stream(price).sum());
                }else {
                    ans[1] = Arrays.stream(price).sum();
                }
                ans[0] = join;
            }
        }

        return ans;
    }

    public static void dfs(int depth) {
        if (depth == tmp.length) {
            discount.add(Arrays.copyOf(tmp, tmp.length));
            return;
        }

        for (int d : data) {
            tmp[depth] += d;
            dfs(depth+1);
            tmp[depth] -= d;
        }
    }

    public static void main(String[] args) {
        int[][] users = {{40, 10000}, {25, 10000}};
        int[] emoticons = {7000, 9000};
        System.out.println(Arrays.toString(solution(users, emoticons)));
    }
}
